### ISSUE TITLE
@jpike88 https://www.npmjs.com/package/@types/node-pushnotifications

### ISSUE THREAD

**Issue by deivi98** [appfeel/node-pushnotifications#193]
@jpike88 https://www.npmjs.com/package/@types/node-pushnotifications

_Originally posted by @alex-friedl in https://github.com/appfeel/node-pushnotifications/issues/120#issuecomment-549041155_

**Comment by miqmago** [comment]
I've created a new branch that's completely rewritten into typescript.
It's a big refactoring and needs some strong review and test, not only myself.
I think it can be the next big release for this repo and make it easier to maintain.

@alex-friedl since you've been contributing a lot to this repo, I've sent you a message, maybe we can do some thinking together.

**Comment by miqmago** [comment]
@vitalyster would be awesome if you could explain little bit your feedback, did you take a look at the branch? Is it crazy? Please let me know how would you like to collaborate!

**Comment by vitalyster** [comment]
@miqmago I'm pretty sure it is much easier and less error prone to maintain existing type definitions but not to do a complete rewrite which will add zero value but require a lot of work

**Comment by miqmago** [comment]
@vitalyster thank for your comments. Completely agree on that it is easier to keep code as it is right now. On the other side, in my opinion typescript is becoming the new standard. Also in fact typescript is something that does not require any rewrite, one could change the name of the files and bum! they are in typescript ☺️

The big effort is on the testing area, which is something that anyway needs someone to take a look every once and then.

On the other side, right now the control of type definitions is over this repo which is not really desirable...

Let's see if there is more time available, I'm planning to use this module for long term and it is interesting to do things that are maintainable in a long term vision. The only drawback is that it will not be immediate. Maybe we will have to survive with outdated type definitions except if there is someone who maintains them...

**Comment by simon-abbott** [comment]
> I'm pretty sure it is much easier and less error prone to maintain existing type definitions but not to do a complete rewrite which will add zero value but require a lot of work

I disagree. Rewriting in TS adds a lot of value: it ensures that the type definitions will always be 100% in sync with the code. That is a massive improvement in type correctness and safety. Right now the latest version of `@types/node-pushnotifications` is v3.1.1, but the latest version of this package is v6.0.0. That is three whole major versions behind. If the types were included in the project itself then the types would always be up-to-date. Writing in TS directly also prevents type mismatches -- when using separate `@types/` packages (or even handwritten `.d.ts` files) it is possible for the types to not actually match the real code. This is an even bigger issue than missing types as it gives a false sense of security, since the TS build passes, but when you go to actually run it you can get unexpected errors.

Maintaining a TypeScript codebase is not much more work than raw JS, with several massive advantages. I urge you to reconsider your stance on this as it would be a huge developer experience improvement.
